server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java

```diff
@@ -0,0 +1,189 @@
+    }
+
+    private static double tile2lat(final double y, final double tiles) {
+        double n = Math.PI - (2.0 * Math.PI * y) / tiles;
+        return Math.toDegrees(Math.atan(Math.sinh(n)));
+    }
+
+    /**
+     * Parse a precision that can be expressed as an integer or a distance measure like "1km", "10m".
+     *
+     * The precision is expressed as a zoom level between 0 and MAX_ZOOM.
+     *
+     * @param parser {@link XContentParser} to parse the value from
+     * @return int representing precision
+     */
+    public static int parsePrecision(XContentParser parser) throws IOException, ElasticsearchParseException {
+        XContentParser.Token token = parser.currentToken();
+        if (token.equals(XContentParser.Token.VALUE_NUMBER)) {
+            return XContentMapValues.nodeIntegerValue(parser.intValue());
+        } else {
+            String precision = parser.text();
+            try {
+                // we want to treat simple integer strings as precision levels, not distances
+                return XContentMapValues.nodeIntegerValue(precision);
+            } catch (NumberFormatException e) {
+                // try to parse as a distance value
+                final int parsedPrecision = GeoUtils.quadTreeLevelsForPrecision(precision);
+                try {
+                    return checkPrecisionRange(parsedPrecision);
+                } catch (IllegalArgumentException e2) {
+                    // this happens when distance too small, so precision > max.
+                    // We'd like to see the original string
+                    throw new IllegalArgumentException("precision too high [" + precision + "]", e2);
+                }
+            }
+        }
+    }
+
+    public static int checkPrecisionRange(int precision) {
+        if (precision < 0 || precision > MAX_ZOOM) {
+            throw new IllegalArgumentException("Invalid quadkey aggregation precision of " +
+                precision + ". Must be between 0 and " + MAX_ZOOM + ".");
+        }
+        return precision;
+    }
+
+    /**
+     * Encode lon/lat to the quadkey based long format.
+     * The resulting hash contains interleaved tile X and Y coordinates.
+     * The precision itself is also encoded as a few high bits.
+     */
+    public static long longEncode(double longitude, double latitude, int precision) {
+        // Adapted from https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#Java
+
+        // How many tiles in X and in Y
+        final int tiles = 1 << checkPrecisionRange(precision);
+        final double lon = normalizeLon(longitude);
+        final double lat = normalizeLat(latitude);
+
+        int xTile = (int) Math.floor((lon + 180) / 360 * tiles);
+        int yTile = (int) Math.floor(
+            (1 - Math.log(
+                Math.tan(Math.toRadians(lat)) + 1 / Math.cos(Math.toRadians(lat))
+            ) / Math.PI) / 2 * tiles);
+        if (xTile < 0)
+            xTile = 0;
```

Comment:
Style nit: I think we try to avoid conditionals without braces, too easy to accidentally introduce bugs if you're not paying attention when touching the code.

Could replace with `xTile = Math.min(xTile, 0)`, if you're looking for a concise alternative.  Probably similar for the rest too.

Reply:
thx, refactored.